### PR TITLE
Driver-only ECC: all three top-level modules

### LINKED ISSUE #7272: Driver-only ECC: all three top-level modules
Context: see #6839 

Once TL.ecdsa, TL.ecdh and TL.ecjpake have been achieved though previous tasks, it's time to turn to the full TL: all three modules accelerated at once. This will materialize in a new `all.sh` component `component_test_psa_crypto_config_accel_ecc_use_psa` (plus its reference companion).

Once this component is added, the previous components `accel_{ecdsa,ecdh,ecjapke}_use_psa` can probably be removed, or at least changed to all use the same reference component. We should strive for a balance between test coverage and use of CI resources which tend to be limited.

### PR DESCRIPTION
**Pull request by valeriosetti**
This is the closing PR of the accelerated EC based algs: all 3 algorithms (ECDSA, ECDH, ECJPAKE) are accelerated at the same time.
Resolves #7272 
Depends on #7312 

## Gatekeeper checklist

- [x] **changelog** not required: changelog for every single accelerated component was already added in previous PRs
- [x] **backport** not required: enhancement
- [X] **tests** provided

### CODE REVIEW

File: library/psa_crypto.c
Comment (mpg) [inline library/psa_crypto.c:380]: Ah, I'm not so sure about that, I was chatting with Ronald and it turns out using `MBEDLTS_xxx_C` macros in PSA files is causing issues with the repo split, so we shouldn't do it. Actually I'm no longer convinced what we did in #7192 is correct. I'll think about it a bit more and come back to you about it.

**Review by mpg — COMMENTED** [review]
Just a first pass, making myself familiar with things.

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh:2191]: I totally agree. It seems that I erroneously cancelled to many lines here :/

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh:2173]: Yes, this [edit here](https://github.com/Mbed-TLS/mbedtls/pull/7321#discussion_r1144668109)  actually refers to this part :)

File: library/psa_crypto.c
Comment (mpg) [reply library/psa_crypto.c:380]: After further chatting with Ronald, it is acceptable for the time being to use `MBEDTLS_xxx_C` feature macros in those parts of PSA that don't have driver support yet, which is the case for ECC derivation, so we're on the right side of the line here after all. Phew!

File: library/ssl_misc.h
Comment (mpg) [inline library/ssl_misc.h:759]: I seem to remember @AndrzejKurek pointing out that guards were inconsistent between `ssl_misc.h` and `ssl_tls.c` (I think) but I can't find the comment again.

Anyway, I have a look and it seems weird to have `|| ECDSA` here, this is about ECDH so the guard should be just ECDH. Turns out if we look a few lines up, there's a comment literally saying this does not seem correct!

Since we are touching this, we should make things right. It's not good having guards with comments saying "this does not seem correct but turned out to be necessary to avoid failures". Can you try removing `|| defined(MBEDTLS_PK_CAN_ECDSA_SOME)` then we'll see what breaks and try to fix it the right way this time?

Comment (valeriosetti) [reply library/ssl_misc.h:759]: > I seem to remember @AndrzejKurek pointing out that guards were inconsistent between ssl_misc.h and ssl_tls.c (I think) but I can't find the comment again.

Yes it was in the previous PR.

>Turns out if we look a few lines up, there's a comment literally saying this does not seem correct! 

TBH I was going to remove that guard as well at first, but then I saw the comment and I didn't want to dig too much into it, so I left it there. I will try to remove it see how it goes. There are 2 options in my opinion:

- find a better description for this guard, or
- the guard is not necessary at all

Comment (mpg) [reply library/ssl_misc.h:759]: From a logical perspective, the `#if` condition really does not look correct: why would we need to have an ECDH context when only ECDSA is enabled? This means either something's wrong in another part of the code and we should fix it, or I'm missing some subtlety and the comment above should explain those subtleties so that we understand what's going on next time we look at this part of the code :)

File: tests/scripts/analyze_outcomes.py
Comment (AndrzejKurek) [inline tests/scripts/analyze_outcomes.py:210]: Shouldn't removed components be also removed from here?

File: tests/scripts/all.sh
Comment (AndrzejKurek) [inline tests/scripts/all.sh:2205]: I'm sorry, but this comment is really hard to read, even though I understand what's the reason behind it. This is probably because of the double negation that reads strange. I would suggest spliting this sentence and explaining that `loc_accel_list` is accelerated by both the test driver library and the main library (by running both sed commands on this variable), and that we want SHA_XXX only to be accelerated on the driver side to allow the driver code to use it, but not to use the SHA_XXX drivers directly by the library.

File: tests/scripts/analyze_outcomes.py
Comment (valeriosetti) [reply tests/scripts/analyze_outcomes.py:210]: Ops, yes, of course, you are totally right.

File: library/ssl_misc.h
Comment (valeriosetti) [reply library/ssl_misc.h]: I [removed that guard](https://github.com/Mbed-TLS/mbedtls/pull/7321/commits/2032a1d5d76e18d5b57ccabfa75d8c5be0d1100e) and run `analyze_outcomes.py` and everything went smoothly so it seems that it was not necessary after all

File: library/ssl_misc.h
Comment (mpg) [inline library/ssl_misc.h:759]: Thanks for fixing this. However, please remove the comment above as it doesn't make sense any more :)

Btw, there's one more occurrence in `ssl.h` around line 47, with a similar comment that we should be able to remove now as well.

Comment (valeriosetti) [reply library/ssl_misc.h:759]: Yes, of course, sorry for that

File: tests/scripts/all.sh
Comment (mpg) [reply tests/scripts/all.sh:2205]: Note: I intend to rework this at some point in the (hopefully near) future, as we're starting to have quite a lot of duplicated code and comments between the various `accel` components. When I do, I'll try to make this clearer at the same time. (But since I don't have a timeframe for that, we should still try to make things readable in the meantime.)

One thing we could do that might improve the situation a bit without requiring too many changes to replace this code block with something like:
```
# Things we wanted supported in libtestdriver1, but not accelerated in the main library:
# SHA-1 and all SHA-2 variants, as they are used by ECDSA deterministic.
loc_extra_list="ALG_SHA1 ALG_SHA_224 ALG_SHA_256 ALG_SHA_384 ALG_SHA_512"
loc_accel_flags=$( echo "$loc_accel_list $loc_extra_list" | sed 's/[^ ]* */-DLIBTESTDRIVER1_MBEDTLS_PSA_ACCEL_&/g' )
make ... # [as before]
```

Wdyt?

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:2286]: (Commenting not on this line but on the component.) Do we still need this component, though? I tend to think we're OK with three components for individual modules (`accel_pake`, `accel_ecdsa` and `accel_ecdh`) not using PSA, and then the component added by this PR accelerating all 3 and `USE_PSA`.

Do you see any gain in having all 3 accelerated without `USE_PSA` in addition?

Comment (valeriosetti) [reply tests/scripts/all.sh:2286]: You are right, we don't need this since `component_test_psa_crypto_config_[accel/reference]_all_ec_algs_use_psa()`  has better coverage

File: tests/scripts/all.sh
Comment (AndrzejKurek) [reply tests/scripts/all.sh:2240]: `# Things we want accelerated in libtestdriver1, but not in the main library:` sounds a bit better in my opinion, but doesn't convey all the information that the original comment had. 
I would say that this is minor, since your're planning to rework it anyway, so let's not focus too much on that.

**Review by mpg — DISMISSED** [review]
LGTM, thanks!

File: tests/scripts/all.sh
Comment (AndrzejKurek) [inline tests/scripts/all.sh:1]: Very minor: commit 4dfdab4a821d3cdc30b title too long.

**Review by AndrzejKurek — DISMISSED** [review]

File: library/ssl_tls12_client.c
Comment (mpg) [reply library/ssl_tls12_client.c:834]: Could perhaps be fixed in #7103 as I think we'll want to rebase it once this one has been merged (currently is has a commit that cherry-picks one of the fixes here).

**Comment by mpg** [comment]
The CI has a couple of failures [in `make test` in the new component](https://ci.staging.trustedfirmware.org/blue/rest/organizations/jenkins/pipelines/mbed-tls-pr-head/branches/PR-7321-head/runs/12/nodes/427/steps/2935/log/?start=0), that would appear to be memory errors during return code translation, in `test_suite_x509write`, probably the "Certificate Request check Server5 ECDSA, key_usage" test case:
```
[2023-03-27T08:44:11.361Z] ==3139==ERROR: AddressSanitizer: global-buffer-overflow on address 0x000000614f38 at pc 0x000000519013 bp 0x7ffe45905c90 sp 0x7ffe45905c80
[2023-03-27T08:44:11.361Z] READ of size 4 at 0x000000614f38 thread T0
[2023-03-27T08:44:11.361Z]     #0 0x519012 in psa_status_to_mbedtls (/var/lib/build/tests/test_suite_x509write+0x519012)
[2023-03-27T08:44:11.361Z]     #1 0x4efe6d in ecdsa_sign_wrap (/var/lib/build/tests/test_suite_x509write+0x4efe6d)
[2023-03-27T08:44:11.361Z]     #2 0x4ec089 in mbedtls_pk_sign_restartable (/var/lib/build/tests/test_suite_x509write+0x4ec089)
[2023-03-27T08:44:11.361Z]     #3 0x4ec146 in mbedtls_pk_sign (/var/lib/build/tests/test_suite_x509write+0x4ec146)
[2023-03-27T08:44:11.361Z]     #4 0x4a9b2b in mbedtls_x509write_csr_der (/var/lib/build/tests/test_suite_x509write+0x4a9b2b)
[2023-03-27T08:44:11.361Z]     #5 0x4aa05e in mbedtls_x509write_csr_pem (/var/lib/build/tests/test_suite_x509write+0x4aa05e)
[2023-03-27T08:44:11.361Z]     #6 0x405599 in test_x509_csr_check (/var/lib/build/tests/test_suite_x509write+0x405599)
[2023-03-27T08:44:11.361Z]     #7 0x405b4f in test_x509_csr_check_wrapper (/var/lib/build/tests/test_suite_x509write+0x405b4f)
[2023-03-27T08:44:11.361Z]     #8 0x40aaf7 in execute_tests (/var/lib/build/tests/test_suite_x509write+0x40aaf7)
[2023-03-27T08:44:11.361Z]     #9 0x4034cb in main (/var/lib/build/tests/test_suite_x509write+0x4034cb)
[2023-03-27T08:44:11.361Z]     #10 0x7f32cd01383f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
[2023-03-27T08:44:11.361Z]     #11 0x4045d8 in _start (/var/lib/build/tests/test_suite_x509write+0x4045d8)
[2023-03-27T08:44:11.361Z] 
[2023-03-27T08:44:11.361Z] 0x000000614f38 is located 40 bytes to the left of global variable 'psa_to_pk_rsa_errors' defined in 'psa_util.c:68:28' (0x614f60) of size 64
[2023-03-27T08:44:11.361Z] 0x000000614f38 is located 0 bytes to the right of global variable 'psa_to_pk_ecdsa_errors' defined in 'psa_util.c:83:28' (0x614f00) of size 56
[2023-03-27T08:44:11.361Z] SUMMARY: AddressSanitizer: global-buffer-overflow ??:0 psa_status_to_mbedtls
```

**Comment by valeriosetti** [comment]
@mpg @AndrzejKurek Sorry for the force push, that was not intentional (initially): there was a GitHub issue, but I thought it was due to something I did locally, so I attempted a rebase. Long story short, it was not my problem, but the history is gone anyway :/

BTW today's work starts from https://github.com/Mbed-TLS/mbedtls/pull/7321/commits/96da36804b9c86dc6ad47dab156102582d554601

**Comment by AndrzejKurek** [comment]
There's one additional thing to fix:
`psa_status_to_mbedtls` should not divide the incomming error numbers by 2. Suggested fix:
```
int psa_status_to_mbedtls(psa_status_t status,
                          const mbedtls_error_pair_t *local_translations,
                          size_t local_errors_num,
                          int (*fallback_f)(psa_status_t))
{
    for (size_t i = 0; i < local_errors_num; i++) {
        if (status == local_translations[i].psa_status) {
            return local_translations[i].mbedtls_error;
        }
    }
    return fallback_f(status);
}
```
That's a remnant of the change to how the errors were stored that happened during the review.

**Review by AndrzejKurek — DISMISSED** [review]

**Review by mpg — APPROVED** [review]
LGTM

File: tests/scripts/all.sh
Comment (AndrzejKurek) [inline tests/scripts/all.sh:2197]: Minor: There's no mention of the input argument.

**Review by AndrzejKurek — APPROVED** [review]
Looks good, minor improvements (probably) for a separate PR suggested, thanks!

**Comment by valeriosetti** [comment]
@AndrzejKurek thanks for the review and feedback!

> minor improvements (probably) for a separate PR suggested

There is #7103 which is waiting for this one to be merged, so in case there is no other change required for this PR, I will apply your suggestions in that PR

**Review by mpg — APPROVED** [review]
Approving the rebase. I've double-checked that the 3 commits that have been removed had all landed development in the meantime.

**Pull request merged**
